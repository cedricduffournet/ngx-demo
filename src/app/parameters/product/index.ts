export * from './parameter-product.component';
