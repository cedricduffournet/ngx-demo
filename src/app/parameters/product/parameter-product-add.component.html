<app-product-add-view></app-product-add-view>
