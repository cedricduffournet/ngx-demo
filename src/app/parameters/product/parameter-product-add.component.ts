import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-parameter-product-add',
  templateUrl: './parameter-product-add.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ParameterProductAddComponent implements OnInit {
  constructor() {}

  ngOnInit() {}
}
