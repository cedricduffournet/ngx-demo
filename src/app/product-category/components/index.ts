export * from './product-category-add/product-category-add.component';
export * from './product-category-update/product-category-update.component';
export * from './product-category-delete/product-category-delete.component';
export * from './product-category-form/product-category-form.component';
export * from './product-category-item/product-category-item.component';
export * from './product-category-items/product-category-items.component';
