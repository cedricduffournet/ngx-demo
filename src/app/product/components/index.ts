export * from './product-add/product-add.component';
export * from './product-update/product-update.component';
export * from './product-delete/product-delete.component';
export * from './product-form/product-form.component';
export * from './product-item/product-item.component';
export * from './product-items/product-items.component';
export * from './product-details/product-details.component';
