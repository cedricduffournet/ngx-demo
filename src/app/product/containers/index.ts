export * from './product-add-view/product-add-view.component';
export * from './product-update-modal/product-update-modal.component';
export * from './product-delete-modal/product-delete-modal.component';
export * from './product-list-view/product-list-view.component';
export * from './product-selected/product-selected.component';
export * from './product-details-view/product-details-view.component';
