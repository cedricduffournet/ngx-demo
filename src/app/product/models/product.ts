import { schema } from 'normalizr';
import { ProductCategory, productCategorySchema } from '@app/product-category/models/product-category';

export interface Product {
  id: number;
  name: string;
  categories: ProductCategory[];
}

export const productSchema = new schema.Entity('products', {
  categories: [productCategorySchema]
});
