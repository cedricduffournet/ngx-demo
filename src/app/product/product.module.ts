import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { PipesModule } from '@app/shared/pipes/pipe.module';

import { ProductCategoryModule } from '@app/product-category/product-category.module';
import { ProductEffects } from '@app/product/state/effects';
import { ProductFacade } from '@app/product/state/product.facade';
import {
  ProductAddComponent,
  ProductUpdateComponent,
  ProductDeleteComponent,
  ProductFormComponent,
  ProductItemComponent,
  ProductItemsComponent,
  ProductDetailsComponent
} from '@app/product/components';
import {
  ProductAddViewComponent,
  ProductUpdateModalComponent,
  ProductDeleteModalComponent,
  ProductListViewComponent,
  ProductSelectedComponent,
  ProductDetailsViewComponent
} from '@app/product/containers';
import { reducers } from '@app/product/state/reducers';
import { ProductGuard } from '@app/product/services/product.guard';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    PipesModule,
    ProductCategoryModule,
    EffectsModule.forFeature([ProductEffects]),
    StoreModule.forFeature('products', reducers)
  ],
  declarations: [
    ProductListViewComponent,
    ProductAddComponent,
    ProductUpdateComponent,
    ProductDeleteComponent,
    ProductFormComponent,
    ProductItemComponent,
    ProductItemsComponent,
    ProductAddViewComponent,
    ProductUpdateModalComponent,
    ProductDeleteModalComponent,
    ProductDetailsComponent,
    ProductSelectedComponent,
    ProductDetailsViewComponent
  ],
  entryComponents: [
    ProductUpdateModalComponent,
    ProductDeleteModalComponent
  ],
  exports: [ProductListViewComponent, ProductAddViewComponent, ProductDetailsViewComponent],
  providers: [ProductFacade, ProductGuard]
})
export class ProductModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
