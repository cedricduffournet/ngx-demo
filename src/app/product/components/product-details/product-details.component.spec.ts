import { TestBed, ComponentFixture, async } from '@angular/core/testing';
import { By } from '@angular/platform-browser';

import { TranslateModule } from '@ngx-translate/core';

import { ActionsItemsModule } from '@app/shared/actions-item';
import { PipesModule } from '@app/shared/pipes/pipe.module';
import { ProductDetailsComponent } from '@app/product/components';
import { ProductFull } from '@app/product/models/product';

describe('ProductDetails', () => {
  let fixture: ComponentFixture<ProductDetailsComponent>;
  let component: ProductDetailsComponent;
  const product: ProductFull = {
    id: 1,
    name: 'testname',
    description: 'Description',
    priceAmount: 1000,
    categories: [
      {
        id: 1,
        name: 'Category 1'
      },
      {
        id: 2,
        name: 'Category 2'
      }
    ]
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [TranslateModule.forRoot(), ActionsItemsModule, PipesModule],
      declarations: [ProductDetailsComponent]
    });

    fixture = TestBed.createComponent(ProductDetailsComponent);
    component = fixture.componentInstance;
    component.authorization = {
      create: false,
      update: false,
      delete: false
    };
    component.product = product;
  });

  it('should be created', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should have description', () => {
    fixture.detectChanges();
    const elem = fixture.nativeElement.querySelector('p');
    expect(elem.textContent).toEqual('Description');
  });

  it('should have product name', () => {
    fixture.detectChanges();
    const elem = fixture.nativeElement.querySelector('h1');
    expect(elem.textContent).toEqual('testname');
  });

  it('should have 2 category', () => {
    fixture.detectChanges();
    const elem = fixture.nativeElement.querySelectorAll('.category');
    expect(elem.length).toBe(2);
  });

  it('should not have edit button', () => {
    fixture.detectChanges();
    const elem = fixture.nativeElement.querySelector('button.btn-edit');
    expect(elem).toBeFalsy();
  });

  it('should have edit button', () => {
    component.authorization = {
      create: false,
      update: true,
      delete: false
    };
    fixture.detectChanges();
    const elem = fixture.nativeElement.querySelector('button.btn-edit');
    expect(elem).toBeTruthy();
  });

  it('should emit update event', () => {
    spyOn(component.update, 'emit');

    component.authorization = {
      create: false,
      update: true,
      delete: false
    };
    fixture.detectChanges();
    const elem = fixture.debugElement.query(By.css('app-actions-item'));
    elem.triggerEventHandler('edit', {});

    expect(component.update.emit).toHaveBeenCalled();
  });

  it('should not have delete button', () => {
    fixture.detectChanges();
    const elem = fixture.nativeElement.querySelector('button.btn-delete');
    expect(elem).toBeFalsy();
  });

  it('should have delete button', () => {
    component.authorization = {
      create: false,
      update: false,
      delete: true
    };
    fixture.detectChanges();
    const elem = fixture.nativeElement.querySelector('button.btn-delete');
    expect(elem).toBeTruthy();
  });

  it('should emit delete event', () => {
    spyOn(component.delete, 'emit');

    component.authorization = {
      create: false,
      update: false,
      delete: true
    };
    fixture.detectChanges();
    const elem = fixture.debugElement.query(By.css('app-actions-item'));
    elem.triggerEventHandler('delete', {});

    expect(component.delete.emit).toHaveBeenCalled();
  });
});
