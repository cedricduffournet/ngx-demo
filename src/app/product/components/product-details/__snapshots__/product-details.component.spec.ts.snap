// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails should be created 1`] = `
<app-product-details
  authorization={[Function Object]}
  delete={[Function EventEmitter]}
  product={[Function Object]}
  update={[Function EventEmitter]}
>
  <div
    class="product-detail"
  >
    <h1>
      testname
    </h1>
    <h2>
      1000 €
    </h2>
    <p>
      Description
    </p>
    
    <div
      class="category-list"
    >
      <span>
        <strong>
          Categories:
        </strong>
      </span>
      
      <span
        class="category"
      >
         Category 1  
      </span>
      <span
        class="category"
      >
         Category 2  
      </span>
    </div>
    <div
      class="float-right actions"
    >
      <app-actions-item
        ng-reflect-with-delete="false"
        ng-reflect-with-edit="false"
      >
        <div
          class="button-validation-action"
        >
          
          
        </div>
      </app-actions-item>
    </div>
  </div>
</app-product-details>
`;
